doc/adr/0008-jetstream-kv.md

```diff
@@ -0,0 +1,326 @@
+# 8. JetStream based Key-Value Stores
+
+|Metadata|Value|
+|--------|-----|
+|Date    |2021-06-30|
+|Author  |@ripienaar|
+|Status  |Partially Implemented|
+|Tags    |jetstream, client, kv|
+
+## Context
+
+This document describes a design and initial implementation of a JetStream backed key-value store. The initial implementation
+is available in the CLI as `nats kv` with an initial reference client package implementation in `jsm.go` repository to support
+the CLI.
+
+This document aims to guide client developers in implementing this feature in the language clients we maintain.
+
+## Overview
+
+We intend to hit a basic initial feature set as below, with some future facing goals as indicated:
+
+Initial feature list:
+
+ * Multiple named buckets full of keys with n historical values kept per value
+ * Put and Get of `string(k)=string(v)` values
+ * Put only if the sequence of the last value for a key matches an expected sequence
+ * Historic values can be kept per key, setable on creation of the bucket
+ * Key deletes preserves history
+ * Keys can be expired from the bucket based on a TTL
+ * Watching a specific key or the entire bucket for live updates
+ * Encoders and Decoders that transforms both keys and values
+ * A read-cache that builds an in-memory cache for fast reads
+ * read-after-write safety unless read replicas used
+ * Valid keys are `\A[-/_a-zA-Z0-9]+\z` after encoding
+ * Valid buckets are `^[a-zA-Z0-9_-]+$`
+ * Custom Stream Names and Stream ingest subjects to cater for different domains, mirrors and imports
+ * Key starting with `_kv` is reserved for internal use
+ * CLI tool to manage the system as part of `nats`, compatible with client implementations
+ * Read-Only and Read-Write interfaces
+
+Planned features:
+
+ * Ability to read from regional read replicas maintained by JS Mirroring (planned)
+ * A read-only client can talk directly, and only, to a regional read replica
+ * Discoverability of read replicas per cluster based on in-bucket configuration values
+ * Standard encoders/decoders to deliver zero-trust end to end encryption, perhaps using nkeys
+ * Pluggable storage backends
+
+## Data Types
+
+### Result
+
+This is the result made available over watchers, `Get()` etc. All backends must return an implementation providing at least this information.
+
+```go
+type Result interface {
+	// Bucket is the bucket the data was loaded from
+	Bucket() string
+	// Key is the key that was retrieved
+	Key() string
+	// Value is the retrieved value
+	Value() string
```

Comment:
can be whatever the language wants as stated - if the nats.go implementation likes []byte go for it.  The interface here is basically a rough intent rather than the rule of the law. 

As long as the things are interpreted in the end `foo` as `foo` its fine.